Function call type checking for operations

## Description

When calling operations, the argument types are of course checked against the declared type. But which actually is the declared type if there are multiple definition of an operation given by the following?

```
interface Eff[A] {
  def op(x: A): A
}
```

Consider this object:

```
def impl: Eff[Int] = new Eff[Int] {
  def op(x: Int) = x / 1
}
```

When calling `impl.op("hello")`, the function type for the symbol of `op` is looked up. However, the function type is annotated as `[A](x: A): A` _once_. Thus, when checking the function call, the call is only ever checked with respect to the original declared type and not all potentially existing types of concrete implementations.

One solution could be to annotate each concrete type during typechecking the implementation of `Eff`, but since we only have exactly one symbol for all possible implementations of `op`, updating the type of `op` overwrites the previous annotation.

Note that this is not an issue exclusive to objects, but to also for handlers.

```
try { eff.op("hello") }
with Eff[Int] { def op(x) = resume(x / 1) }
```

Both of these examples compile and yield `NaN`.

## Idea

We need to have unique symbols for each operation implementation and then annotate the specific type for each implementation to this unique symbol.

## Relevant

- https://github.com/effekt-lang/effekt/blob/5dffce2c7f61b03c5038ddfe9d3c1f0fc0ae2ae1/effekt/shared/src/main/scala/effekt/Typer.scala#L364
  We need to bind this type to the operation here: `Context.bind(op.symbol.asOperation, <>)`. This overwrites previous annotations, though.

- https://github.com/effekt-lang/effekt/blob/5dffce2c7f61b03c5038ddfe9d3c1f0fc0ae2ae1/effekt/shared/src/main/scala/effekt/Typer.scala#L176-L177
- https://github.com/effekt-lang/effekt/blob/5dffce2c7f61b03c5038ddfe9d3c1f0fc0ae2ae1/effekt/shared/src/main/scala/effekt/Typer.scala#L910-L912

--- a/effekt/jvm/src/test/scala/effekt/ConstraintTests.scala
+++ b/effekt/jvm/src/test/scala/effekt/ConstraintTests.scala
@@ -33,7 +33,7 @@ abstract class ConstraintTests extends munit.FunSuite {
 
 
   def freshTypeVar(name: String) =
-    scope.fresh(TypeParam(Name.local(name)), NoSource)
+    scope.freshTypeVar(TypeParam(Name.local(name)), NoSource)
 
   def freshCaptVar(name: String) =
     scope.freshCaptVar(CaptUnificationVar.VariableInstantiation(CaptureParam(Name.local(name)), NoSource))

--- a/effekt/shared/src/main/scala/effekt/Typer.scala
+++ b/effekt/shared/src/main/scala/effekt/Typer.scala
@@ -950,13 +950,41 @@ object Typer extends Phase[NameResolved, Typechecked] {
 
     val Result(recvTpe, recvEffs) = checkExprAsBlock(receiver, None)
 
-    val interface = recvTpe.asInterfaceType.typeConstructor
+    val interface = recvTpe.asInterfaceType
     // filter out operations that do not fit the receiver
-    val candidates = methods.filter(op => op.interface == interface)
+    val candidates = methods.filter(op => op.interface == interface.typeConstructor)
 
     val (successes, errors) = tryEach(candidates) { op =>
       val (funTpe, capture) = findFunctionTypeFor(op)
-      checkCallTo(call, op.name.name, funTpe, targs, vargs, bargs, expected)
+
+      // 1) check arity of explicitly provided type arguments
+
+      if (targs.nonEmpty && targs.size != funTpe.tparams.size)
+        Context.abort(s"Wrong number of type arguments, given ${targs.size} but expected ${funTpe.tparams.size}")
+
+      // 2)
+      // args present: check prefix against receiver
+      (targs zip interface.args).foreach { case (manual, inferred) =>
+        matchExpected(inferred, manual)
+      }
+
+      // args missing: synthesize args from receiver and unification variables (e.g. [Int, String, ?A, ?B])
+      def fillInTypeArguments(tparams: List[TypeParam], interfaceArgs: List[ValueType]): List[ValueType] =
+        (tparams, interfaceArgs) match {
+          // we have an argument, provided by the interface: use it
+          case (param :: params, arg :: args) => arg :: fillInTypeArguments(params, args)
+          // we don't have an argument, create fresh unification variable
+          case (param :: params, Nil) => ValueTypeRef(Context.freshTypeVar(param, call)) :: fillInTypeArguments(params, Nil)
+          case (Nil, _) => Nil
+        }
+
+      val synthTargs = if targs.nonEmpty then targs else fillInTypeArguments(funTpe.tparams, interface.args)
+
+      // TODO maybe type checking all calls should have a similar structure like above:
+      //   1. make up and annotate unification variables, if no type arguments are there
+      //   2. type check call with either existing or made up type arguments
+
+      checkCallTo(call, op.name.name, funTpe, synthTargs, vargs, bargs, expected)
     }
     resolveOverload(id, List(successes), errors)
   }
@@ -1341,7 +1369,7 @@ trait TyperOps extends ContextOps { self: Context =>
    * allow to save a copy of the current state.
    */
   private[typer] val unification = new Unification(using this)
-  export unification.{ requireSubtype, requireSubregion, join, instantiate, freshCaptVar, without, requireSubregionWithout }
+  export unification.{ requireSubtype, requireSubregion, join, instantiate, freshCaptVar, freshTypeVar, without, requireSubregionWithout }
 
   // opens a fresh unification scope
   private[typer] def withUnificationScope[T](additional: List[CaptUnificationVar])(block: => T): T = {

--- a/effekt/shared/src/main/scala/effekt/typer/Unification.scala
+++ b/effekt/shared/src/main/scala/effekt/typer/Unification.scala
@@ -49,7 +49,7 @@ class Unification(using C: ErrorReporter) extends TypeUnifier, TypeMerger, TypeI
 
   // Creating fresh unification variables
   // ------------------------------------
-  def fresh(underlying: TypeVar.TypeParam, call: source.Tree): UnificationVar = scope match {
+  def freshTypeVar(underlying: TypeVar.TypeParam, call: source.Tree): UnificationVar = scope match {
     case GlobalScope => sys error "Cannot add unification variables to global scope"
     case s : LocalScope =>
       val x = new UnificationVar(underlying, call)
@@ -200,7 +200,7 @@ class Unification(using C: ErrorReporter) extends TypeUnifier, TypeMerger, TypeI
 
     val typeRigids =
       if (targs.size == tparams.size) targs
-      else tparams map { t => ValueTypeRef(fresh(t, position)) }
+      else tparams map { t => ValueTypeRef(freshTypeVar(t, position)) }
 
     if (cparams.size != (bparams.size + eff.canonical.size)) {
       sys error pp"Capture param count ${cparams.size} is not equal to bparam ${bparams.size} + controleffects ${eff.canonical.size}.\n  ${tpe}"

--- a/examples/neg/issue373/existentials.check
+++ b/examples/neg/issue373/existentials.check
@@ -1,0 +1,6 @@
+[error] examples/neg/issue373/existentials.effekt:7:46: Expected Int but got String.
+  with eff: Eff[Int] { def op[C](x) = resume(x ++ "") }
+                                             ^^^^^^^
+[error] examples/neg/issue373/existentials.effekt:7:46: Expected String but got C.
+  with eff: Eff[Int] { def op[C](x) = resume(x ++ "") }
+                                             ^

--- a/examples/neg/issue373/existentials.effekt
+++ b/examples/neg/issue373/existentials.effekt
@@ -1,0 +1,7 @@
+interface Eff[A] {
+  def op[B](x: B): A
+}
+
+def main() =
+  try { println(eff.op("hello")) }
+  with eff: Eff[Int] { def op[C](x) = resume(x ++ "") }

--- a/examples/neg/issue373/existentials_wrong_arity.check
+++ b/examples/neg/issue373/existentials_wrong_arity.check
@@ -1,0 +1,3 @@
+[error] examples/neg/issue373/existentials_wrong_arity.effekt:6:17: Wrong number of type arguments, given 1 but expected 2
+  try { println(eff.op[Int]("hello")) }
+                ^^^^^^^^^^^^^^^^^^^^

--- a/examples/neg/issue373/existentials_wrong_arity.effekt
+++ b/examples/neg/issue373/existentials_wrong_arity.effekt
@@ -1,0 +1,7 @@
+interface Eff[A] {
+  def op[B](x: A): A
+}
+
+def main() =
+  try { println(eff.op[Int]("hello")) }
+  with eff: Eff[Int] { def op[B](x) = resume(x / 1) }

--- a/examples/neg/issue373/handler_correct_annotation.check
+++ b/examples/neg/issue373/handler_correct_annotation.check
@@ -1,0 +1,3 @@
+[error] examples/neg/issue373/handler_correct_annotation.effekt:6:29: Expected Int but got String.
+  try { println(eff.op[Int]("hello")) }
+                            ^^^^^^^

--- a/examples/neg/issue373/handler_correct_annotation.effekt
+++ b/examples/neg/issue373/handler_correct_annotation.effekt
@@ -1,0 +1,7 @@
+interface Eff[A] {
+  def op(x: A): A
+}
+
+def main() =
+  try { println(eff.op[Int]("hello")) }
+  with eff: Eff[Int] { def op(x) = resume(x / 1) }

--- a/examples/neg/issue373/handler_no_annotation.check
+++ b/examples/neg/issue373/handler_no_annotation.check
@@ -1,0 +1,3 @@
+[error] examples/neg/issue373/handler_no_annotation.effekt:6:24: Expected Int but got String.
+  try { println(eff.op("hello")) }
+                       ^^^^^^^

--- a/examples/neg/issue373/handler_no_annotation.effekt
+++ b/examples/neg/issue373/handler_no_annotation.effekt
@@ -1,0 +1,7 @@
+interface Eff[A] {
+  def op(x: A): A
+}
+
+def main() =
+  try { println(eff.op("hello")) }
+  with eff: Eff[Int] { def op(x) = resume(x / 1) }

--- a/examples/neg/issue373/object_no_annotation.check
+++ b/examples/neg/issue373/object_no_annotation.check
@@ -1,0 +1,3 @@
+[error] examples/neg/issue373/object_no_annotation.effekt:9:29: Expected Int but got String.
+def main() = println(eff.op("hello"))
+                            ^^^^^^^

--- a/examples/neg/issue373/object_no_annotation.effekt
+++ b/examples/neg/issue373/object_no_annotation.effekt
@@ -1,0 +1,9 @@
+interface Eff[A] {
+  def op(x: A): A
+}
+
+def eff = new Eff[Int] {
+  def op(x) = x / 1
+}
+
+def main() = println(eff.op("hello"))

--- a/examples/neg/issue373/wrong_type_argument.check
+++ b/examples/neg/issue373/wrong_type_argument.check
@@ -1,0 +1,6 @@
+[error] examples/neg/issue373/wrong_type_argument.effekt:6:17: Expected Boolean but got Int.
+  try { println(eff.op[Boolean]("hello")) }
+                ^^^^^^^^^^^^^^^^^^^^^^^^
+[error] examples/neg/issue373/wrong_type_argument.effekt:6:33: Expected Boolean but got String.
+  try { println(eff.op[Boolean]("hello")) }
+                                ^^^^^^^

--- a/examples/neg/issue373/wrong_type_argument.effekt
+++ b/examples/neg/issue373/wrong_type_argument.effekt
@@ -1,0 +1,7 @@
+interface Eff[A] {
+  def op(x: A): A
+}
+
+def main() =
+  try { println(eff.op[Boolean]("hello")) }
+  with eff: Eff[Int] { def op(x) = resume(x / 1) }

--- a/examples/neg/issue373/wrong_type_argument_arity.check
+++ b/examples/neg/issue373/wrong_type_argument_arity.check
@@ -1,0 +1,3 @@
+[error] examples/neg/issue373/wrong_type_argument_arity.effekt:6:17: Wrong number of type arguments, given 2 but expected 1
+  try { println(eff.op[Boolean, Int]("hello")) }
+                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

--- a/examples/neg/issue373/wrong_type_argument_arity.effekt
+++ b/examples/neg/issue373/wrong_type_argument_arity.effekt
@@ -1,0 +1,7 @@
+interface Eff[A] {
+  def op(x: A): A
+}
+
+def main() =
+  try { println(eff.op[Boolean, Int]("hello")) }
+  with eff: Eff[Int] { def op(x) = resume(x / 1) }

